avm/src/vm/memory.rs

```diff
@@ -1,654 +1,390 @@
+      let mut orig_arr_addrs: Vec<usize> = vec![a];
+      let mut orig_arr_copies: Vec<Vec<(usize, i64)>> = vec![orig.read_fractal(orig_addr).to_vec().clone()];
+      while check_idx < orig_arr_addrs.len() {
+        let arr = &orig_arr_copies[check_idx];
+        let l = arr.len();
+        drop(arr);
+        for i in 0..l {
+          let other_arr_idx = orig_arr_copies[check_idx][i].0.clone();
+          if other_arr_idx < std::usize::MAX {
+            if !orig_arr_addrs.contains(&other_arr_idx) {
+              orig_arr_addrs.push(other_arr_idx);
+              orig_arr_copies.push(orig.mems[other_arr_idx].clone());
+            }
+          }
+        }
+        check_idx = check_idx + 1;
       }
-    } else {
-      // We can see a difference between the normal and unsafe forms of reading these integers in
-      // benchmarking
-      unsafe {
-        let mem_ptr = self.mem.as_mut_ptr();
-        *mem_ptr.add(addr as usize) = payload;
+      // Next, get the current size of the destination mems vector to use as the offset to add to
+      // the index of the copied arrays, updating their interior references, if any, in the process
+      let dest_offset = dest.mems.len();
+      for i in 0..orig_arr_copies.len() {
+        let arr = &mut orig_arr_copies[i];
+        for j in 0..arr.len() {
+          let (a, b) = arr[j];
+          if a < std::usize::MAX {
+            for k in 0..orig_arr_addrs.len() {
+              if orig_arr_addrs[k] == a {
+                arr[j] = (dest_offset + k, b);
+              }
+            }
+          }
+        }
       }
+      dest.mems.append(&mut orig_arr_copies);
+      // Finally, set the destination address to point at the original, main nested array
+      dest.set_addr(dest_addr, dest_offset, std::usize::MAX);
     }
   }
 
-  // new fractal from mem
-  pub fn write_fractal_mem(self: &mut HandlerMemory, addr: i64, payload: &[i64]) {
-    let arr = HandlerMemory {
-      gmem: self.gmem,
-      mem: payload.to_vec(),
-      fractal_mem: vec![],
-      either_mem: vec![-1; payload.len()],
-      registers_ish: vec![0; payload.len()],
-      is_fixed: false,
-      closure_args: Vec::new(),
-    };
-    self.write_fractal(addr, arr);
-  }
-
-  pub fn write_fractal(self: &mut HandlerMemory, addr: i64, payload: HandlerMemory) {
-    if addr < 0 && !self.is_neg_addr_gmem(addr) {
-      let closure = &mut self.closure_args[0];
-      return closure.write_fractal(addr - CLOSURE_ARG_MEM_START, payload);
-    } else {
-      let idx = self.fractal_mem.len() as i64;
-      self.either_mem[addr as usize] = idx;
-      self.fractal_mem.push(payload);
-      self.registers_ish.push(0);
+  pub fn dupe(self: &mut HandlerMemory, orig_addr: i64, dest_addr: i64) {
+    // This *should be possible with something like this:
+    // HandlerMemory::transfer(self, orig_addr, self, dest_addr);
+    // But Rust's borrow checker doesn't like it, so we basically have to replicate the code here
```

Comment:
ugh..

Reply:
I know. I don't even understand why it's a problem. :/ I hope it's fixed in a later version of Rust.